File: exercises/small_problems/medium_2.rb
#1 Longest Sentence
def longest_sentence(str)
  sentences = str.scan(/[^.?!]+[.?!]/).map { |sentence|
    sentence.gsub("\n"," ").strip
  }
  sentences.map { |sentence| sentence.split(" ").count }.max
end

#2 Now I Know My ABCs
ABC = {"B" => "O", "X" => "K", "D" => "Q", "C" => "P", "N" => "A",
"G" => "T", "R" => "E", "F" => "S", "J" => "W", "H" => "U", "V" => "I",
"L" => "Y", "Z" => "M" }

def block_word?(str)
  dumpster = []
  str.upcase.chars.each do |char|
    if ABC.keys.include?(char)
      dumpster << char << ABC[char]
    elsif ABC.values.include?(char)
      dumpster << char << ABC.key(char)
    else
      dumpster << nil
    end
  end
   if dumpster.uniq.length == dumpster.length && !dumpster.include?(nil)
    return true
  end
  false
end

#3 Lettercase Percentage Ratio
LOWERCASE = ('a'..'z').to_a
UPPERCASE = ('A'..'Z').to_a

def letter_percentages(str)
  totals = {}
  lower = []
  upper = []
  neither = []
  str.chars.each do |char|
    if LOWERCASE.include?(char)
      lower << char
    elsif UPPERCASE.include?(char)
      upper << char
    else
      neither << char
    end
  end
  total_count = (lower.size + upper.size + neither.size).to_f
  totals["lowercase"] = (lower.size / total_count) * 100
  totals["uppercase"] = (upper.size / total_count) * 100
  totals["neither"] = (neither.size / total_count) * 100
  totals
end

#4 Matching Parentheses?
def balanced?(str)
  parens = 0

  str.chars.each do |char|
    parens += 1 if char == "("
    parens -= 1 if char == ")"
    break if parens < 0
  end
  parens.zero?
end

#5 Triangle Sides
def triangle(a, b, c)
  sides = [a,b,c]
  max = sides.max

  if sides.include?(0) || max >= (sides.reduce(:+) - max)
    :invalid
  elsif a == b && b == c
    :equilateral
  elsif a == b || b == c || a == c
    :isosceles
  else
    :scalene
  end
end

#6 Tri-Angles
def triangle(a, b, c)
  angles = [a,b,c]

  if angles.reduce(:+) != 180 || angles.include?(0)
    :invalid
  elsif angles.include?(90)
    :right
  elsif angles.any? { |angle| angle > 90 }
    :obtuse
  else
    :acute
  end
end

#7 Unlucky Days
def friday_13th(yr)
  month = 1
  count = 0
  loop do
    count += 1 if Time.new(yr, month, 13).friday?
    month += 1
    break if month > 12
  end
  count
end

#8 Next Featured Number Higher than a Given Value
def featured(num)
  test = num + 1
  loop do
    if test.odd? && (test % 7 == 0) && test.to_s.chars.size == test.to_s.chars.uniq.size
      break
    else
      test += 1
    end
  end
  test
end

#9 Bubble Sort
def bubble_sort!(array)
  loop do
    swapped = false
    1.upto(array.size - 1) do |i|
      next if array[i-1] <= array[i]
      array[i-1], array[i] = array[i], array[i-1]
      swapped = true
    end
    break unless swapped
  end
end

#10 Sum Square - Square Sum
def sum_square_difference(num)
(1..num).reduce(:+) ** 2 - (1..num).reduce(0) { |sum, int| sum + int **2}
end

File: js101_foundations/exercises/medium_2.js
/* eslint-disable */
// Lettercase Percentage Ratio
const letterPercentages = str => {
  let lowercase = str.match(/[a-z]/g) || [];
  let uppercase = str.match(/[A-Z]/g) || [];
  let neither = str.match(/[^a-zA-Z]/g) || [];

  lowercase = (lowercase.length / str.length * 100).toFixed(2);
  uppercase = (uppercase.length / str.length * 100).toFixed(2);
  neither = (neither.length / str.length * 100).toFixed(2);

  return { lowercase: String(lowercase), uppercase: String(uppercase), neither: String(neither) };
}

// Triangle Sides
const triangle = (a, b, c) => {
  let array = [a,b,c].sort((a,b) => a - b);

  if ((array[0] + array[1]) < array[2] || array[0] <= 0) {
    return "invalid";
  }

  if (a === b && b === c) {
    return "equilateral";
  } else if (a === b || b === c || a === c) {
    return "isosceles";
  } else {
    return "scalene";
  }
}

// Tri-Angles
const triangle = (a, b, c) => {
  let array = [a,b,c].sort((a,b) => a - b);

  if (a + b + c !== 180 || array[0] <= 0) {
    return "invalid";
  }

  if (array.includes(90)) {
    return "right";
  } else if (array[2] > 90) {
    return "obtuse";
  } else {
    return "acute";
  }
}

// Unlucky Days
const fridayThe13ths = year => {
  let days = [];
  for(let i = 0; i <= 11; i++) {
    days.push(new Date(year, i, 13).getDay());
  }
  return days.filter(num => num === 5).length;
}

// Next Featured Number Higher than a Given Value
const uniqueDigits = num => {
  let counts = {};
  let array = String(num).split("");

  for (let i = 0; i < array.length; i++) {
    if (counts[array[i]]) {
      return false;
    }

    counts[array[i]] = true;
  }
  return true;
}

const featuredNum = num => {
  return num % 7 === 0 && num % 2 === 1 && uniqueDigits(num);
}

const featured = num => {
  for (let i = num + 1; i <= 9876543201; i++) {
    if (featuredNum(i)) {
      return i;
    }
  }

  return "There is no possible number that fulfills those requirements.";
}

// Sum Square - Square Sum
const sumSquareDifference = count => {
  let sum = 0;
  let squaredSum = 0;
  for(let i = 1; i <= count; i++) {
    sum += i;
    squaredSum += i ** 2;
  }
  return sum ** 2 - squaredSum;
}

// Bubble Sort
const bubbleSort = array => {
  while (true) {
    let swapped = false;
    for (let idx = 1; idx < array.length; idx++) {
      if (array[idx - 1] <= array[idx]) continue;
      [array[idx - 1], array[idx]] = [array[idx], array[idx - 1]];
      swapped = true;
    }
    if (!swapped) break;
  }
}

// Longest Sentence
const longestSentence => text => {
  let sentences = text.match(/\w.*?[.!?]/g);

  let longest = sentences.reduce((longest,sentence) => {
    let length = sentence.split(/\s/).length;
    if (length > longest.length) {
      return { text: sentence, length: length };
    } else {
      return longest;
    }
  }, { text: "", length: 0 });

  console.log(longest.text + "\n");
  console.log("The longest sentence has " + longest.length + " words.\n");
}